Improve model configuration flexibility and fallback behavior

robottwo opened the pull request:
This MR enhances the model configuration system in ragit with two key improvements:

1. Custom model configuration sources - Adds support for loading model configurations from:

- RAGIT_MODEL_CONFIG environment variable
- ~/.config/ragit/models.json file
- Existing per-repository configuration

2. Automatic fallback to lowest-cost model - Ensures ragit always uses a valid model:

- When the default model is unavailable in the user's configuration
- Selects the lowest-cost model as a fallback
- Displays a warning message when fallback occurs

3. Override defaults using ~/.config/ragit/{api.json, search.json, build.json} when init'ing a new rag.

- Allows you to set different parameters as "standard" across the types of data you are working with



Benefits

- Users can maintain consistent model configurations and RAG settings across repositories
- Improves robustness by ensuring ragit always uses a valid model
- Provides clear feedback when fallback occurs

API keys are not copied to the models by default, which allows the user to share rag-indexed directories with other users, and those other users will use their own API keys (falling back to their ~/.config/ragit/models.json keys) rather than the API key of the user who created the RAG.

baehyunsol commented:
Thanks so much for a contribution! I really like the idea of adding a global configuration, and fallbacks.

I'd like to merge your PR, but could you add some changes before I merge your PR?

1. There's `ragit_api::get_model_by_name(models: &[Model], name: &str) -> Result<Model, Error>` function in `ragit_api` crate. Could you use this function instead of iterating the models? It returns `Err` if there's no match or multiple matches.
2. Could you replace `join3(&home_dir, ".config/ragit", "models.json")?;` in `src/index.rs` with `join4(&home_dir, ".config", "ragit", "models.json")?;` You can import this from `ragit_fs` if it's not already imported.
3. Could you move `test_models_init.py` to `tests/models_init.py`? If you don't mind, I want to add this to my CI pipeline. Also, I want to change a few lines of code in this file, so that it's more consistent with other tests.
4. Your PR is likely to be included in 0.3.2 release, and I'll copy and paste your release note. If you want something else to be included in the release note, please leave a comment.

Thanks a lot!

robottwo commented:
No problem ... updated!

robottwo commented:
I also added another patch to this MR, which basically does the same thing for the other config files.

FWIW, the motivation for me was that I want to default the rag search to return more than 3 results :)

baehyunsol commented:
Nice! Could you edit your PR message? I'll copy-paste that to the release note, and I want it to mention the new commit you just made.

robottwo commented:
Thanks. I fixed a bug to make sure it handles partial config files correctly in the top level, and I updated the PR message at the top of this thread.
